### PR TITLE
Parser performance improvements

### PR DESCRIPTION
**Pull request by jokva**
Depends on https://github.com/OPM/opm-parser/pull/705

This patch set addresses some of the performance shortcomings and inefficiencies in opm-parsers, mostly tracked down using valgrind (callgrind). Each individual commit addresses some specific aspect. A short summary:

* string_view is a std::string interface compatible (not completely, but on a by-need basis) view into a (sub)string to avoid copying too many strings.
* boost algorithms have been replaced by faster hand-rolled (locale-ignoring) versions.
* some containers take size hints and preallocates memory.
* some algorithmic improvements and heuristics applied.

### CODE REVIEW

**Comment by andlaus** [comment]
that's pretty cool! What's the performance improvement you see for real decks, though? (I'm thinking about Norne and "Model 2". A ballpark measure is enough I suppose.)

**Comment by jokva** [comment]
On my machine, Norne is parsed in ~5.3s with master. With this patch set applied it takes ~1.8s.

**Comment by andlaus** [comment]
Oh, only a speedup of factor 2.5? that's definitely not worth the trouble. (just kidding ;)

**Comment by jokva** [comment]
I just checked Model 2. Master takes ~11.92s, this patch ~3.3s.

**Comment by joakim-hove** [comment]
Wow - that was impressive!

**Comment by jokva** [comment]
I'll rebase this on top of master on Monday (which will resolve the merge conflict). It can obviously be reviewed now, but if you want to test it you might have to rebase locally.

**Comment by atgeirr** [comment]
Great work! Just a simple question: why is string_view needed, would it not be possible to use const string references (or pointers)? Also: could not string_view::substr() return string_view instances?

**Comment by jokva** [comment]
>Great work! Just a simple question: why is string_view needed, would it not be possible to use const string references (or pointers)?

string_view is needed because the RawRecord (used to build deck) stores an entire input set in one string (whereas previously it stored  them in a deque of smaller strings copied out of the larger input string which was a massive performance killer), meaning we'd need to store the string ref + offset + size anyway, which is what string_view does.

It's certainly possible, but then it's implementing string_view anyway. 

>Also: could not string_view::substr() return string_view instances? If it did I would see greater benefit of it.

It could, but I chose to mimic std::string's interface in order to have all other code using substr "just work"™. If there is need for a string_view substring to view feature then substrview or substr_view is ok.

>I guess the name of the class is chosen to align with possible future standardized features, but I fear this may be more of a problem than a benefit.

Partially - ideally this implementation goes away once we move on to a standardised version. I don't like the UpperCaseNames at all, and especially not since this has nothing to do with opm or reservoir logic, but is strictly a memory/implementation/optimisation thing.

**Comment by jokva** [comment]
I've rebased this onto master.

**Comment by jokva** [comment]
I have now rebased this onto master.

**Comment by acelster** [comment]
Great work, Joergen! Also nice to see you are still contributing to OPM.
Glad I got you involved. ;-) Anne

File: opm/parser/eclipse/CMakeLists.txt [hunk 1]
```diff
@@ -287,9 +288,12 @@ EclipseState/Tables/ColumnSchema.hpp
 EclipseState/Tables/TableEnums.hpp
 EclipseState/Tables/TableSchema.hpp
 EclipseState/Tables/TableIndex.hpp
-Utility/Functional.hpp)
+#
+Utility/Functional.hpp
+Utility/Stringview.hpp
+Utility/EquilWrapper.hpp)
```
Comment (joakim-hove) [inline opm/parser/eclipse/CMakeLists.txt]: Not really part of this PR; but can you move `EquilWrapper` to `EclipseState/InitConfig`?

File: opm/parser/eclipse/Deck/DeckRecord.cpp [hunk 2]
```diff
@@ -29,6 +29,10 @@
 namespace Opm {
 
 
+    DeckRecord::DeckRecord( size_t size ) {
```
Comment (joakim-hove) [inline opm/parser/eclipse/Deck/DeckRecord.cpp]: It certainly makes a lot of sense to reserve the size of each individual `DeckItem` - but reserving the number of `DeckItem` instances in a `DeckRecord` probably does not make much sense performance wise?

File: opm/parser/eclipse/CMakeLists.txt (hunk 1 above)
Comment (jokva) [inline opm/parser/eclipse/CMakeLists.txt]: This should be gone and is a rebase bug. i'll get rid of it.

File: opm/parser/eclipse/Deck/DeckRecord.cpp (hunk 2 above)
Comment (jokva) [inline opm/parser/eclipse/Deck/DeckRecord.cpp]: It's pretty significant actually, I seem to recall 8-10% improvement. Mostly because reallocating here is expensive because it invokes clone.

**Comment by joakim-hove** [comment]
Maybe the `Rawxxxx` classes should be renamed `InputXxxx`?

**Comment by jokva** [comment]
It feels slightly more natural I suppose, I have no strong objections. Not sure if it's worth the effort as I'm not sure how much more this part will be touched, but it can surely be done. It really isn't that clear that 'raw' largely means the tokenising phase.

**Comment by joakim-hove** [comment]
Good - then I think we should do 's/Raw/Input/'.
11. mars 2016 12:42 skrev "jokva" <notifications@github.com> følgende:

File: opm/parser/eclipse/RawDeck/RawRecord.hpp [hunk 3]
```diff
@@ -34,32 +37,27 @@ namespace Opm {
...
-        virtual ~RawRecord();
-        void dump() const;
+
+       void dump() const;
 
     private:
         std::string m_sanitizedRecordString;
-        std::deque<std::string> m_recordItems;
+        std::deque< string_view > m_recordItems;
+        std::vector< std::string > expanded_items;
```
Comment (joakim-hove) [inline opm/parser/eclipse/RawDeck/RawRecord.hpp]: I guess these strings are only to ensure that views which are `push_front()`'ed have something to view? In that case e.g. `string_storage` would be a better name.

File: opm/parser/eclipse/RawDeck/RawRecord.hpp [hunk 4]
```diff
@@ -34,32 +37,27 @@ namespace Opm {
...
 
-        std::string pop_front();
+        string_view pop_front();
         void push_front(std::string token);
         size_t size() const;
 
         const std::string& getRecordString() const;
-        const std::string& getItem(size_t index) const;
+        string_view getItem(size_t index) const;
         const std::string& getFileName() const;
```
Comment (joakim-hove) [inline opm/parser/eclipse/RawDeck/RawRecord.hpp]: If moving the too-many-items handling from `ParserRecord::parse()` to `ParserKeyword::parse()` we can get rid of the `filename` and `keyword` members of the `RawRecord` class. Maybe not much performance to gain - but slightly cleaner.

File: opm/parser/eclipse/RawDeck/RawRecord.hpp (hunk 3 above)
Comment (jokva) [inline opm/parser/eclipse/RawDeck/RawRecord.hpp]: Kind-of. The expanded_items stores the physical strings generated by *-expansion, which isn't stored in the sanitised string (it actually is, we could push back the same view to the star value multiple times). The vector is always empathy in the abscence of * in the input file.

**Comment by joakim-hove** [comment]
OK  - this is very good; except the minor rename+++ points mentioned I am ready to merge this.

**Comment by jokva** [comment]
I'll look at this first thing Monday.

**Comment by joakim-hove** [comment]
Just one thought: Would it be possible to re*use* the `RawRecord/RawKeyword` instance(s) - so that it is not necessary to reallocate the `std::vector<string_view>` member of the `RawRecord` for every record?

**Comment by joakim-hove** [comment]
Just a quick question: The fixes are all in the `input -> Deck` stage - no fixes in the `Deck -> EclipseState`. Is that because that is where you focused your effort, or is it because the `Deck -> EclipseState` phase was anyway fast?

**Comment by jokva** [comment]
> Just one thought: Would it be possible to reuse the RawRecord/RawKeyword instance(s) - so that it is not necessary to reallocate the std::vector<string_view> member of the RawRecord for every record?

I suppose you *can*, but I sincerely doubt it's worth it. Primarily for code clarity reasons (as we need the Raw* to stick around until their corresponding Deck* has been constructed), but also because the 1-allocation-per isn't that big of a deal. Still something and it might be worth investigating surely.

The bigger costs came from repeated allocation or resizing. With a shared instance we'd probably let vector or deque reallocate itself until it reaches the longest record or item was reached, then reuse the memory. Since a lot of records are really small they don't really trigger any *re-allocations* so they can be considered cheap (no copying or moving of elements). I can think a bit about it, but my gut feeling tells me it's hardly any benefit in this.

> Just a quick question: The fixes are all in the input -> Deck stage - no fixes in the Deck -> EclipseState. Is that because that is where you focused your effort, or is it because the Deck -> EclipseState phase was anyway fast?

When I started, Input -> Deck accounted for approx. 90% of the run time, and I found the pain points where part-of-execution-time was too high (i.e. anything that isn't directly related to either AST management or string-to-int/double conversion. Now that number is closer to 80%, and while I'm sure there are plenty of inefficiencies in Deck -> EclipseState it simply hasn't been that much to gain from improving it yet.
